### PR TITLE
Additional DAGMC test for lost particles

### PR DESCRIPTION
**Pull request by pshriwise**
This is one I've been meaning to add for a while. I was recently prompted by some discussion in #1926.

I've taken the DAGMC pincell model we use in a few other tests and removed 100 random triangles from the surfaces of the cladding volume to ensure that we'd encounter some situations in which an intersection with the particle's DAGMC cell cannot be found. 

One of the main purposes of this test is to ensure that lost particles are caught correctly in DAGMC models not only when the DAGMC model is the root universe, but also when DAGMC universes are used to fill other cells.

@gridley  raise a very good question in #1926 about why the `DAGCell::contains` method handles a missing intersection differently by marking the particle as lost in the method rather than a like `CSGCell`, which returns an invalid surface index and an infinite intersection distance. 

For DAGMC universes that are used as the root universe of the geometry, we can expect that a finite intersection distance is _always_ found. If this were the only way to use DAGMC universes, we could return the same behavior as `CSGSurfaces` without a problem. 

But in the case that a DAGMC universe is used as a universe filling a cell or lattice, an infinite intersection distance isn't valid _unless_ the particle is in the implicit complement. If the particle is in the implicit complement, we need to allow for the case that the particle has left all explicitly defined regions of the DAGMC model and will travel up to the boundary of the containing cell/lattice cell to continue transport. If we were to use the same behavior as the `CSGCell` in this case and return an infinite intersection distance, a missed intersection in an explicitly defined DAGMC volume that should result in a lost particle will allow the particle to jump to the boundary of the containing cell through all other volumes in the DAGMC model because of how our surface intersection logic works -- the particle would intersect with the nearest surface higher up in the geometry hierarchy. So, because there isn't a way (that I can see) to capture this nuance for DAGMC universes, we need to report the particle as lost in the `DAGCell::distance` method.*

That was really lengthy... I can provide a diagram if requested by the reviewer.

I also updated `lattice.py` to use `openmc.UniverseBase` instead of `openmc.Universe`  because it was breaking the `Lattice.get_all_universes` method when DAGMC universes are present. This update is now tested in some of the lines of code in the new test here.

* Noting here that while writing this I realized that this doesn't cover a case where we are in the implicit complement of a DAGMC universe filling a cell/lattice cell and we _should_ get a finite intersection distance but don't (due to watertightness issues). I'm not sure how we'd detect that offhand and will give it some thought.

### CODE REVIEW

**Comment by pshriwise** [comment]
> Makes sense to me! CSG surfaces cannot have holes, while mesh surfaces may. Cool.
> 
> (or can they? The new torus surfaces _can_ have a hole so we may be missing some weird particle leak edge case there)

Not quite sure what you mean here. As in there's a known particle leakage problem for torus surfaces?

> I'll give this some time so you can figure out a good solution to that last issue.

Thought about this some more this morning. Regardless of whether the fill universe is composed of CSG or DAGMC cells, the scenario in which a particle doesn't intersect correctly with its current cell and immediately "jumps" to the boundary of the universe's containing cell is something we generally can't handle with our current approach. Because this isn't a problem that's unique to DAGMC universes, I think we can proceed with the PR and address the [new issue](https://github.com/openmc-dev/openmc/issues/2438) I made related to this topic separately.

**Comment by pshriwise** [comment]
> Should we not mark the particle as lost in dagmc if we're in plotting mode, then? This PR would be the place to add that behavior.

Yeah, that would be my suggestion for now. Should be a one-liner in that block that marks particles as lost in the `DAGCell`.

**Comment by gridley** [comment]
I would review this, but am not familiar enough with DAGMC to provide any useful feedback! Generally, the changes look good to me.

**Comment by shimwell** [comment]
I was just skimming through the open PRs and noticed that a single conflict has crept into this one. I don't trust myself with the C++ enough to resolve it sorry

**Comment by pshriwise** [comment]
Thanks @shimwell! Just resolved that conflict 👍🏻

File: tests/unit_tests/dagmc/test_lost_particles.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    cladding.add_nuclide('Zr92', 0.0072758)
+    cladding.add_nuclide('Zr94', 0.0073734)
+    cladding.add_nuclide('Zr96', 0.0011879)
+
+    water = openmc.Material(name='water')
+    water.set_density('g/cc', 0.740582)
+    water.add_nuclide('H1', 0.049457)
+    water.add_nuclide('O16', 0.024672)
+    water.add_nuclide('B10', 8.0042e-06)
+    water.add_nuclide('B11', 3.2218e-05)
```
Comment (paulromano) [inline tests/unit_tests/dagmc/test_lost_particles.py:40]: Can you cut these down to a single nuclide each? It's not important for the sake of the test and results in more time in the test just loading cross sections

File: tests/unit_tests/dagmc/test_lost_particles.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    left = openmc.XPlane(x0=-pitch[0], name='left', boundary_type='reflective')
+    right = openmc.XPlane(x0=pitch[0], name='right', boundary_type='reflective')
+    front = openmc.YPlane(y0=-pitch[1], name='front', boundary_type='reflective')
+    back = openmc.YPlane(y0=pitch[1], name='back', boundary_type='reflective')
+    # clip the DAGMC geometry at +/- 10 cm w/ CSG planes
+    bottom = openmc.ZPlane(z0=-10.0, name='bottom', boundary_type='reflective')
+    top = openmc.ZPlane(z0=10.0, name='top', boundary_type='reflective')
+
+    bounding_region = +left & -right & +front & -back & +bottom & -top
+    bounding_cell = openmc.Cell(fill=lattice, region=bounding_region)
```
Comment (paulromano) [inline tests/unit_tests/dagmc/test_lost_particles.py:67]:
```suggestion
    # clip the DAGMC geometry at +/- 10 cm w/ CSG planes
    rpp = openmc.model.RectangularParallelepiped(
        -pitch[0], pitch[0], -pitch[1], pitch[1], -10.0, 10.0, boundary_type='reflective')
    bounding_cell = openmc.Cell(fill=lattice, region=-rpp)
```

File: tests/unit_tests/dagmc/test_lost_particles.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    model.geometry = openmc.Geometry(root=[bounding_cell])
+
+    # settings
+    model.settings.particles = 100
+    model.settings.batches = 10
+    model.settings.inactive = 2
+    model.settings.output = {'summary' : False}
```
Comment (paulromano) [inline tests/unit_tests/dagmc/test_lost_particles.py:75]:
```suggestion
    model.settings.output = {'summary': False}
```

File: tests/unit_tests/dagmc/test_lost_particles.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+pytestmark = pytest.mark.skipif(
+    not openmc.lib._dagmc_enabled(),
+    reason="DAGMC CAD geometry is not enabled.")
+
+
+@pytest.fixture
+def broken_dagmc_model(request):
+    openmc.reset_auto_ids()
+    model = openmc.model.Model()
```
Comment (paulromano) [inline tests/unit_tests/dagmc/test_lost_particles.py:17]:
```suggestion
    model = openmc.Model()
```

**Review by paulromano — CHANGES_REQUESTED** [review]
A few comments here on the test. Otherwise the code looks good to me -- would appreciate a second look from someone more familiar with DAGMC if possible (@shimwell @makeclean)

**Review by paulromano — APPROVED** [review]
Thanks @pshriwise! I'll let this sit over the weekend in case others have thoughts.

**Pull request merged**